bpo-43224: Draft implementation of PEP 646

mrahtz opened the pull request:
~~PEP 646 (Variadic Generics) hasn't yet been completely approved by the steering council, but here's a draft of its implementation so we can get started on code review.~~

Update 2021-01-19: the SC has approved the PEP, so full steam ahead!

TODO:
- [ ] Finish addressing @pradeep90's comments from the previous PR, https://github.com/python/cpython/pull/24527
- [x] Add tests for `typing.get_args` and `typing.get_origin` (rather than just checking `__args__` etc)

A couple of notes below.

@pradeep90 @gvanrossum @Fidget-Spinner

## Starred tuple types

I realised there's a tricky issue with starred tuple types.

If we have e.g. `tuple[int, *tuple[str, bool]]`, the obvious thing to do would be to unpack the inner `tuple` at runtime, resulting in `tuple[int, str, bool]`. This would imply that `iter(tuple[str, bool])` creates an iterator returning `str` then `bool`.

But this would cause problems when using starred tuple types as the type of `*args`, e.g. `def foo(*args: *tuple[str, bool])`:
1. The type of `foo.__annotations__['args']` should definitely not be `tuple[str, bool]`. That would instead imply `def foo(*args: tuple[str, bool])`
2. The current PEP draft states that when a starred object is used as the annotation of `*args`, its iterator should return only a single value. (Or at least, this is the behaviour specified for `*args: *Ts`; @pradeep90, I've just realised, should we update that section of the PEP to make it explicit that that starred tuples can be also used for `*args`?)

I think the simplest solution to both of these problems is to *not* unpack tuple types as runtime, having `*tuple[stuff]` instead creating an iterator that returns only a single item, an instance of a new 'starred tuple'. (Note that such a type does have to exist anyway, in order to represent e.g. `*tuple[int, ...]`.)

For the native `tuple` type, I've tentatively implemented this by adding a new `starred` flag to `gaobject` in `genericaliasobject.c` rather than creating a whole new type. This makes implementation easier, but maybe it's suboptimal considering that `gaobject` is used for lots of things other than `tuple` - feedback appreciated. For `typing.Tuple`, I've implemented a new class in `typing.py` called `StarredTuple`, which should behave in the same way. (This does mean that `*tuple[int] != *Tuple[int]`, but given that `tuple[int] != Tuple[int]`, this seems fine.)

## Implementation of `*tuple`

I'm pretty uncertain about my implementation in `genericaliasobject.c`. In particular:
* I don't know what I'm doing with garbage collection; scrutiny on whether I'm doing it right appreciated.
* I wonder whether there's a way to make it simpler. In `typing.py`, the implementation is as simple as `def __iter__(self): yield StarredTuple(self)`. Is there a way to do something similar in `ga_iter`, without the need to also implement `ga_iternext`?

<!-- issue-number: [bpo-43224](https://bugs.python.org/issue43224) -->
https://bugs.python.org/issue43224
<!-- /issue-number -->

mrahtz committed:
PEP 646: Grammar and compiler changes

mrahtz committed:
PEP 646: Add AST tests for *args annotations

Formatted, the generated AST is as follows.

`def f(*args: *Ts): pass`:

```
(
    "Module",
    [
        (
            "FunctionDef",
            (1, 0, 1, 23),
            "f",
            (
                "arguments",
                [],
                [],
                (
                    "arg",
                    (1, 7, 1, 16),
                    "args",
                    (
                        "Starred",
                        (1, 13, 1, 16),
                        ("Name", (1, 14, 1, 16), "Ts", ("Load",)),
                        ("Load",),
                    ),
                    None,
                ),
                [],
                [],
                None,
                [],
            ),
            [("Pass", (1, 19, 1, 23))],
            [],
            None,
            None,
        )
    ],
    [],
)
```

`def f(*args: *tuple[int, ...]): pass`:

```
(
    "Module",
    [
        (
            "FunctionDef",
            (1, 0, 1, 36),
            "f",
            (
                "arguments",
                [],
                [],
                (
                    "arg",
                    (1, 7, 1, 29),
                    "args",
                    (
                        "Starred",
                        (1, 13, 1, 29),
                        (
                            "Subscript",
                            (1, 14, 1, 29),
                            ("Name", (1, 14, 1, 19), "tuple", ("Load",)),
                            (
                                "Tuple",
                                (1, 20, 1, 28),
                                [
                                    ("Name", (1, 20, 1, 23), "int", ("Load",)),
                                    ("Constant", (1, 25, 1, 28), Ellipsis, None),
                                ],
                                ("Load",),
                            ),
                            ("Load",),
                        ),
                        ("Load",),
                    ),
                    None,
                ),
                [],
                [],
                None,
                [],
            ),
            [("Pass", (1, 32, 1, 36))],
            [],
            None,
            None,
        )
    ],
    [],
)
```

`def f(*args: *tuple[int, *Ts]): pass`:

```
(
    "Module",
    [
        (
            "FunctionDef",
            (1, 0, 1, 36),
            "f",
            (
                "arguments",
                [],
                [],
                (
                    "arg",
                    (1, 7, 1, 29),
                    "args",
                    (
                        "Starred",
                        (1, 13, 1, 29),
                        (
                            "Subscript",
                            (1, 14, 1, 29),
                            ("Name", (1, 14, 1, 19), "tuple", ("Load",)),
                            (
                                "Tuple",
                                (1, 20, 1, 28),
                                [
                                    ("Name", (1, 20, 1, 23), "int", ("Load",)),
                                    (
                                        "Starred",
                                        (1, 25, 1, 28),
                                        ("Name", (1, 26, 1, 28), "Ts", ("Load",)),
                                        ("Load",),
                                    ),
                                ],
                                ("Load",),
                            ),
                            ("Load",),
                        ),
                        ("Load",),
                    ),
                    None,
                ),
                [],
                [],
                None,
                [],
            ),
            [("Pass", (1, 32, 1, 36))],
            [],
            None,
            None,
        )
    ],
    [],
)
```

mrahtz committed:
PEP 646: Add AST tests for return types with unpacking

Formatted, generated AST is as follows.

`def f() -> tuple[*Ts]: pass`:

```
(
    "Module",
    [
        (
            "FunctionDef",
            (1, 0, 1, 27),
            "f",
            ("arguments", [], [], None, [], [], None, []),
            [("Pass", (1, 23, 1, 27))],
            [],
            (
                "Subscript",
                (1, 11, 1, 21),
                ("Name", (1, 11, 1, 16), "tuple", ("Load",)),
                (
                    "Tuple",
                    (1, 17, 1, 20),
                    [
                        (
                            "Starred",
                            (1, 17, 1, 20),
                            ("Name", (1, 18, 1, 20), "Ts", ("Load",)),
                            ("Load",),
                        )
                    ],
                    ("Load",),
                ),
                ("Load",),
            ),
            None,
        )
    ],
    [],
)
```

`def f() -> tuple[int, *Ts]: pass`:

```
(
    "Module",
    [
        (
            "FunctionDef",
            (1, 0, 1, 32),
            "f",
            ("arguments", [], [], None, [], [], None, []),
            [("Pass", (1, 28, 1, 32))],
            [],
            (
                "Subscript",
                (1, 11, 1, 26),
                ("Name", (1, 11, 1, 16), "tuple", ("Load",)),
                (
                    "Tuple",
                    (1, 17, 1, 25),
                    [
                        ("Name", (1, 17, 1, 20), "int", ("Load",)),
                        (
                            "Starred",
                            (1, 22, 1, 25),
                            ("Name", (1, 23, 1, 25), "Ts", ("Load",)),
                            ("Load",),
                        ),
                    ],
                    ("Load",),
                ),
                ("Load",),
            ),
            None,
        )
    ],
    [],
)
```

`def f() -> tuple[int, *tuple[int, ...]]: pass`:

```
(
    "Module",
    [
        (
            "FunctionDef",
            (1, 0, 1, 45),
            "f",
            ("arguments", [], [], None, [], [], None, []),
            [("Pass", (1, 41, 1, 45))],
            [],
            (
                "Subscript",
                (1, 11, 1, 39),
                ("Name", (1, 11, 1, 16), "tuple", ("Load",)),
                (
                    "Tuple",
                    (1, 17, 1, 38),
                    [
                        ("Name", (1, 17, 1, 20), "int", ("Load",)),
                        (
                            "Starred",
                            (1, 22, 1, 38),
                            (
                                "Subscript",
                                (1, 23, 1, 38),
                                ("Name", (1, 23, 1, 28), "tuple", ("Load",)),
                                (
                                    "Tuple",
                                    (1, 29, 1, 37),
                                    [
                                        ("Name", (1, 29, 1, 32), "int", ("Load",)),
                                        ("Constant", (1, 34, 1, 37), Ellipsis, None),
                                    ],
                                    ("Load",),
                                ),
                                ("Load",),
                            ),
                            ("Load",),
                        ),
                    ],
                    ("Load",),
                ),
                ("Load",),
            ),
            None,
        )
    ],
    [],
)
```

mrahtz committed:
PEP 646: Add AST tests for variable annotations with unpacking

Formatted, generated AST is as follows:

`x: tuple[*Ts]`:

```
(
    "Module",
    [
        (
            "AnnAssign",
            (1, 0, 1, 13),
            ("Name", (1, 0, 1, 1), "x", ("Store",)),
            (
                "Subscript",
                (1, 3, 1, 13),
                ("Name", (1, 3, 1, 8), "tuple", ("Load",)),
                (
                    "Tuple",
                    (1, 9, 1, 12),
                    [
                        (
                            "Starred",
                            (1, 9, 1, 12),
                            ("Name", (1, 10, 1, 12), "Ts", ("Load",)),
                            ("Load",),
                        )
                    ],
                    ("Load",),
                ),
                ("Load",),
            ),
            None,
            1,
        )
    ],
    [],
)
```

`x: tuple[int, *Ts]`:

```
(
    "Module",
    [
        (
            "AnnAssign",
            (1, 0, 1, 18),
            ("Name", (1, 0, 1, 1), "x", ("Store",)),
            (
                "Subscript",
                (1, 3, 1, 18),
                ("Name", (1, 3, 1, 8), "tuple", ("Load",)),
                (
                    "Tuple",
                    (1, 9, 1, 17),
                    [
                        ("Name", (1, 9, 1, 12), "int", ("Load",)),
                        (
                            "Starred",
                            (1, 14, 1, 17),
                            ("Name", (1, 15, 1, 17), "Ts", ("Load",)),
                            ("Load",),
                        ),
                    ],
                    ("Load",),
                ),
                ("Load",),
            ),
            None,
            1,
        )
    ],
    [],
)
```

`x: tuple[int, *tuple[str, ...]]`:

```
(
    "Module",
    [
        (
            "AnnAssign",
            (1, 0, 1, 31),
            ("Name", (1, 0, 1, 1), "x", ("Store",)),
            (
                "Subscript",
                (1, 3, 1, 31),
                ("Name", (1, 3, 1, 8), "tuple", ("Load",)),
                (
                    "Tuple",
                    (1, 9, 1, 30),
                    [
                        ("Name", (1, 9, 1, 12), "int", ("Load",)),
                        (
                            "Starred",
                            (1, 14, 1, 30),
                            (
                                "Subscript",
                                (1, 15, 1, 30),
                                ("Name", (1, 15, 1, 20), "tuple", ("Load",)),
                                (
                                    "Tuple",
                                    (1, 21, 1, 29),
                                    [
                                        ("Name", (1, 21, 1, 24), "str", ("Load",)),
                                        ("Constant", (1, 26, 1, 29), Ellipsis, None),
                                    ],
                                    ("Load",),
                                ),
                                ("Load",),
                            ),
                            ("Load",),
                        ),
                    ],
                    ("Load",),
                ),
                ("Load",),
            ),
            None,
            1,
        )
    ],
    [],
)
```

mrahtz committed:
PEP 646: Add support for starring tuple types

E.g. tuple[*tuple[int, str]].

mrahtz committed:
PEP 646: Implement support in typing.py

mrahtz committed:
PEP 646: Add tests for everything in typing.py

mrahtz committed:
PEP 646: Add tests for starring tuple types

mrahtz committed:
PEP 646: Add tests for syntax changes

mrahtz commented on Lib/typing.py:
```diff
@@ -823,6 +874,95 @@ def __init__(self, name, *constraints, bound=None,
+        return self._name
+
+
+class UnpackedTypeVarTuple(_Final, _Immutable, _root=True):
```
@pradeep90 In the previous review, you mentioned you'd prefer to avoid this middleman class, instead for example making it so that starring a TypeVarTuple results in an instance of `Unpack`.

That example solution seems suboptimal to me for two reasons:
1. In a few places in the rest of typing.py, we want to check whether an object is an unpacked TypeVarTuple, and `instance(obj, UnpackedTypeVarTuple)` seems more readable to me than `isinstance(obj, Unpacked)`.
2.  It seems potentially confusing to have `Unpack` both be an operator and an object.

Another option would be something like:

```python
class TypeVarTuple:

    def __init__(self, name, _starred=False, _id=None):
        self.name = name
        self._starred = starred
        self._id = id

    def __iter__(self):
        yield TypeVarTuple(self.name, _starred=True, _id=id(self))

    def __eq__(self, other):
        if self._id is not None and other._id is not None:
            return self._id == other._id
        else:
            return object.__eq__(self, other)  # Not sure whether this is right, but you get the gist

    def __repr__(self):
         return ('*' if self._starred else '') + self.name
```

I lean away from this solution too because I feel like manually managing IDs could lead to subtle bugs, and `isinstance(obj, UnpackedTypeVarTuple)` is still simpler than `isinstance(obj, TypeVarTuple) and obj._starred`, but WDYT?

mrahtz commented on Lib/typing.py:
```diff
@@ -823,6 +874,95 @@ def __init__(self, name, *constraints, bound=None,
             self.__module__ = def_mod
 
 
+class TypeVarTuple(_Final, _Immutable, _root=True):
```
@pradeep90 You mentioned in the previous review you'd prefer to have this inherit from `_TypeVarLike`. I'm not sure I understood your reasoning - could you explain what you meant by "[The flags are things that the typechecker has to deal with](https://github.com/python/cpython/pull/24527/files#r575868925)"?

mrahtz committed:
PEP 646: Add tests for typing.get_{origin,args}

and fix a couple of bugs the tests revealed

Fidget-Spinner commented on Lib/test/test_pep646_syntax.py:
```diff
@@ -0,0 +1,158 @@
+
+def test_main(verbose=False):
+    from test import support
+    from test import test_pep646
```
The CI fails to import this. What's `test_pep646`?

mrahtz replied:
Oops - I renamed this at the last minute. This is meant to be `test_pep646_syntax`. Fixed.

Fidget-Spinner commented on Objects/genericaliasobject.c:
```diff
@@ -626,6 +640,65 @@ static PyNumberMethods ga_as_number = {
+    PyObject_GC_Del(gi);
+}
+
+PyTypeObject Py_GenericAliasIterType = {
```
My C is rusty, but IIRC, this will make `Py_GenericAliasIterType` public to the C API, and we probably don't want that for now until things stabilize a little. Maybe
```suggestion
static PyTypeObject Py_GenericAliasIterType = {
```

mrahtz replied:
Done.

Fidget-Spinner commented on Objects/genericaliasobject.c:
```diff
@@ -626,6 +640,65 @@ static PyNumberMethods ga_as_number = {
+
+    if (gi->exhausted) {
+        gi->obj = NULL;
+        return NULL;
```
I feel like you could potentially be missing a decref here for `gi->obj` since I see it being increfed in `gi_iter`. If so,
```suggestion
        Py_SETREF(gi->obj, NULL);
        return NULL;
```
If I'm just shooting my mouth off, then just please ignore what I said :).

One way to test for refleaks is to run the tests with the refleak hunter (requires a debug build):
```
python_d.exe -m test test_typing -R 3:3
```

mrahtz replied:
Oh, super nice about the refleak hunter - I had no idea that was a thing!

And good catch about the decref too. One thing I'm unsure about is where we should do it, though. Reading https://docs.python.org/3/c-api/gcsupport.html, am I right in thinking that the iterator object `gi` counts as a container type, because it contains references to another object? If so, that same page says "Before fields which refer to other containers are invalidated, PyObject_GC_UnTrack() must be called". That would imply we should be setting `gi->obj` to NULL in the deallocator, in which case we'd also call the setref there?

mrahtz replied:
Friendly poke :)

Fidget-Spinner commented on Objects/genericaliasobject.c:
```diff
@@ -626,6 +640,65 @@ static PyNumberMethods ga_as_number = {
+    starred_tuple->origin = alias->origin;
+    starred_tuple->args = alias->args;
+    starred_tuple->parameters = alias->parameters;
+    starred_tuple->weakreflist = alias->weakreflist;
```
I wonder if ``*starred_tuple = *alias`` works here instead of copying the fields individually. Probably not.

mrahtz replied:
I realised it would be cleaner to construct `starred_tuple` using the `Py_GenericAlias` constructor itself, which makes this a little cleaner anyway :)

Fidget-Spinner commented on Objects/genericaliasobject.c:
```diff
@@ -11,8 +11,17 @@ typedef struct {
+typedef struct {
+    PyObject_HEAD
+    PyObject *obj;  /* Set to NULL when iterator is exhausted */
+    int exhausted;
```
If we're setting `obj` to `NULL` only when it's exhausted, can't we check for that instead of checking for the `exhausted` flag? Then we can get rid of one field and save some memory. Or is there some situation where `obj` is `NULL` even when not exhausted?

mrahtz replied:
Ah, good point. Fixed.

Fidget-Spinner reviewed (commented):
Wow, this is seriously impressive work and a very strong start! I don't have time to review everything right now, so far I've covered `genericaliasobject.c` and `compile.c`. I still have `typing.py` and friends left. I'm hopeless at the parser and grammar so I'll leave those to someone else.

pablogsal commented on Grammar/python.gram:
```diff
@@ -780,7 +787,7 @@ primary[expr_ty]:
 slices[expr_ty]:
     | a=slice !',' { a }
-    | a[asdl_expr_seq*]=','.slice+ [','] { _PyAST_Tuple(a, Load, EXTRA) }
+    | a[asdl_expr_seq*]=','.(slice | starred_expression)+ [','] { _PyAST_Tuple(a, Load, EXTRA) }
```
I would prefer to move `starred_expression` as another option of `slice` as we already have `named_expression` there

mrahtz replied:
I just tried this, but then I get `SyntaxError: can't use starred expression here` for e.g. `a[*b]`.

I think what happens is, if we move `starred_expression` to `slice` (as `| a=starred_expression { a }`), then when we do `a[*b]`, it triggers the _first_ option of `slices`, where the `starred_expression` _doesn't_ get processed by `_PyAST_Tuple`.

I can't see any way to make sure `_PyAST_Tuple` definitely gets called for `starred_expression` without doing _something_ to `slices`, so I think the original attempt with `(slice | starred_expression)` is cleanest. Wdyt?

pablogsal commented on Grammar/python.gram:
```diff
@@ -331,14 +333,19 @@ kwds[arg_ty]: '**' a=param_no_default { a }
 param_no_default[arg_ty]:
     | a=param ',' tc=TYPE_COMMENT? { _PyPegen_add_type_comment_to_arg(p, a, tc) }
     | a=param tc=TYPE_COMMENT? &')' { _PyPegen_add_type_comment_to_arg(p, a, tc) }
+param_no_default_star_annotation[arg_ty]:
```
Why not join these to `param_no_default`? That way we can remove the extra rule in `star_etc`.

mrahtz replied:
Hmm, wouldn't that mess with other things that use `param_no_default` though? E.g. `kwds` uses `param_no_default`, so if we joined these to `param_no_default`, wouldn't we accidentally allow `def foo(***args)`?

mrahtz replied:
Friendly poke :)

mrahtz committed:
PEP 646: Various fixes to genericaliasobject.c

Based on Fidget-Spinner's feedback:
1. Remove the explicit flag for 'exhausted', instead using 'gi->obj == NULL' to tell whether the iterator is exhausted.
2. Construct `starred_tuple` using the `Py_GenericAlias` constructor instead of constructing it manually.
3. Fix reference counting.
4. Make `Py_GenericAliasIterType` static.

mrahtz committed:
PEP 646: Fix test_pep646_syntax import

JelleZijlstra commented on Lib/typing.py:
```diff
@@ -823,6 +874,96 @@ def __init__(self, name, *constraints, bound=None,
+class StarredTuple(_Final, _Immutable, _root=True):
+    """Implementation of starred tuple for older versions of Python.
+
+    A starred tuple is e.g. tuple[*tuple[int, str]]. In newer versions of
```
But this implementation is only used in new versions of Python, so why do we need this?

mrahtz replied:
Ah yeah, thanks for the reminder on this too - I was unsure about this.

The two things I had in mind were:
* It was easier to just put everything in one PR so that when doing code review we could compare between the 'new' and the 'backwards compatible' implementations.
* We probably want _some_ of the 'backwards compatible' implementations even in modern `typing.py` for the sake of code (or maybe even pickle?) compatibility. `Unpack` is probably the most important one, so we can read backwards-compatible code written as `class C(Generic[Unpack[Ts]])`. I'm admittedly less sure about `StarredTuple`; I don't _think_ there's ever a reason to instantiate it explicitly (and in fact maybe it should be private?). In any case, not sure what best practice is in cases like this - @gvanrossum can you comment?

JelleZijlstra commented on Lib/typing.py:
```diff
@@ -823,6 +874,96 @@ def __init__(self, name, *constraints, bound=None,
             self.__module__ = def_mod
 
 
+class TypeVarTuple(_Final, _Immutable, _root=True):
```
I was hoping I could steal a docstring for TypeVarTuple from you for use in typing-extensions :) A docstring should be added here.

mrahtz replied:
Thanks for the reminder :) Done.

JelleZijlstra commented on Lib/typing.py:
```diff
@@ -823,6 +874,96 @@ def __init__(self, name, *constraints, bound=None,
+    def __class_getitem__(cls, item):
+        if isinstance(item, types.GenericAlias) and item.__origin__ is tuple:
+            # tuple[...]
+            return StarredTuple(item)
```
Could this return an instance of Unpack instead so we don't need to add an additional class?

mrahtz replied:
It's definitely an option, but I'd prefer to keep `Unpack` as only implementing the operator, rather implementing both the operator and the wrapper class for unpacked items - in line with the principle of "Each component should only do one thing".

Also, I think `Unpack` could get really messy if we have to merge in the guts of both `StarredTuple` and `UnpackedTypeVarTuple`. And there are a bunch of places where we want to check for an unpacked `TypeVarTuple` but not an unpacked `StarredTuple`, so they would get more complicated too.

Maybe there's context I'm missing though. Is there a big downside (e.g. maintenance burden) to adding another class?

JelleZijlstra commented on Lib/typing.py:
```diff
@@ -823,6 +874,96 @@ def __init__(self, name, *constraints, bound=None,
+        elif isinstance(item, TypeVarTuple):
+            return item._unpacked
+        else:
+            raise TypeError(
```
I'd prefer to not raise an error here, so we don't hobble potential future extensions. For example, I anticipate that we may want to support `Unpack[TypedDict]` to type `**kwargs` in the future.

mrahtz replied:
a) If the concern is future extensions, couldn't we just amend this function when that need arises in the future?

b) Since the result of unpacking depends on what's being unpacked, what would we do instead of raising an error here?

JelleZijlstra reviewed (commented):
Just a couple of things I noticed while working on the typing-extensions implementation.

I'm not going to be the one to merge this, but I think it might make sense to split this PR into separate PRs for the grammar changes and the typing.py changes, since they touch very different parts of the codebase and need review from different people.

mrahtz committed:
PEP 646: Add docstring for TypeVarTuple

kumaraditya303 commented on Lib/test/test_ast.py:
```diff
@@ -13,7 +13,7 @@
 
 def to_tuple(t):
-    if t is None or isinstance(t, (str, int, complex)):
+    if t is None or isinstance(t, (str, int, complex)) or t is Ellipsis:
```
```suggestion
    if t is None or isinstance(t, (str, int, complex, types.EllipsisType)):
```

JelleZijlstra replied:
The current version seems cleaner to me, and easier to port to other Python versions.

mrahtz replied:
In that case, @JelleZijlstra, fair enough; I'll leave it as it is. Thanks anyway, @kumaraditya303.

kumaraditya303 commented on Lib/test/test_genericalias.py:
```diff
@@ -190,6 +196,7 @@ def test_parameters(self):
         D2b = dict[T, T]
         self.assertEqual(D2b.__args__, (T, T))
         self.assertEqual(D2b.__parameters__, (T,))
+
```
```suggestion
```

mrahtz replied:
Ditto above.

kumaraditya303 commented on Lib/test/test_genericalias.py:
```diff
@@ -175,6 +180,7 @@ def test_exposed_type(self):
 
     def test_parameters(self):
         from typing import List, Dict, Callable
+
```
```suggestion
```

mrahtz replied:
I think I'd prefer to keep these line breaks; the test case is so long now that I think it improves readability to have the tests for the different types broken up a bit.

kumaraditya303 commented on Objects/genericaliasobject.c:
```diff
@@ -626,6 +639,50 @@ static PyNumberMethods ga_as_number = {
+    if (gi == NULL)
+        return NULL;
+    gi->obj = self;
+    Py_INCREF(self);
```
```suggestion
    gi->obj = Py_NewRef(self);
```

mrahtz replied:
Nice, thanks!

mrahtz committed:
PEP 646: Swap a Py_INCREF with a Py_NewRef

mrahtz committed:
PEP 646: Add test for assigning to a slice containing a star

blurb-it[bot] committed:
📜🤖 Added by blurb_it.

mrahtz committed:
PEP 646: Add line to Doc/whatsnew

kumaraditya303 commented on Lib/test/test_genericalias.py:
```diff
@@ -212,6 +219,55 @@ def test_parameters(self):
+        with self.assertRaises(TypeError):
+            tuple[*list[int]]
+
+        T0 = tuple[*Ts]
```
Can this be changed to subtests ?

mrahtz replied:
Good point. Done.

kumaraditya303 commented on Objects/genericaliasobject.c:
```diff
@@ -11,8 +11,16 @@ typedef struct {
     PyObject* weakreflist;
+    // Whether we're a starred type, e.g. *tuple[int].
+    // Only supported for `tuple`.
+    int starred;
```
Can you use bool from `stdbool.h` instead of int?

mrahtz replied:
Ooh, I didn't know about `stdbool.h`! Done.

mrahtz committed:
PEP 646: Use stdbool.h; separate test cases
